kotlinx-coroutines-core/common/src/flow/operators/Delay.kt

```diff
@@ -99,6 +105,39 @@ public fun <T> Flow<T>.debounce(timeoutMillis: Long): Flow<T> {
 @FlowPreview
 public fun <T> Flow<T>.debounce(timeout: Duration): Flow<T> = debounce(timeout.toDelayMillis())
 
+private fun <T> Flow<T>.debounceInternal(timeoutMillisSelector: (T) -> Long) : Flow<T> =
+    scopedFlow { downstream ->
+        // Actually Any, KT-30796
+        val values = produce<Any?>(capacity = Channel.CONFLATED) {
+            collect { value -> send(value ?: NULL) }
+        }
+        var lastValue: Any? = null
+        while (lastValue !== DONE) {
+            select<Unit> {
+                // Should be receiveOrClosed when boxing issues are fixed
+                values.onReceiveOrNull {
+                    if (it == null) {
+                        if (lastValue != null) downstream.emit(NULL.unbox(lastValue))
+                        lastValue = DONE
+                    } else {
+                        lastValue = it
+                    }
+                }
+
+                lastValue?.let { value ->
+                    val unboxedValue: T = NULL.unbox(value)
+                    val timeoutMillis = timeoutMillisSelector(unboxedValue)
+                    require(timeoutMillis > 0) { "Debounce timeout should be positive" }
```

Comment:
Can we make 0 valid value as well? In case of using a selector and having different behaviour to different items in the steam, it is quite possible that some of the items would require to have no delay at all. Then we can move this line to `debounce(timeoutMillis: Long)` function to make sure that current behaviour won't change.